packages/react-dom-bindings/src/server/ReactFizzConfigDOM.js

```diff
@@ -2636,36 +2726,73 @@ function pushImg(
       // We must construct a new preload resource
       resumableState.imageResources[key] = PRELOAD_NO_CREDS;
-      resource = [];
-      pushLinkImpl(
-        resource,
-        ({
-          rel: 'preload',
-          as: 'image',
-          // There is a bug in Safari where imageSrcSet is not respected on preload links
-          // so we omit the href here if we have imageSrcSet b/c safari will load the wrong image.
-          // This harms older browers that do not support imageSrcSet by making their preloads not work
-          // but this population is shrinking fast and is already small so we accept this tradeoff.
-          href: srcSet ? undefined : src,
-          imageSrcSet: srcSet,
-          imageSizes: sizes,
-          crossOrigin: props.crossOrigin,
+      const crossOrigin = getCrossOriginString(props.crossOrigin);
+
+      const headers = renderState.headers;
+      if (
+        headers &&
+        headers.remainingCapacity > 0 &&
+        // this is a hueristic similar to capping element preloads to 10 unless explicitly
+        // fetchPriority="high". We use length here which means it will fit fewer images when
+        // the urls are long and more when short. arguably byte size is a better hueristic because
+        // it directly translates to how much we send down before content is actually seen.
+        // We could unify the counts and also make it so the total is tracked regardless of
+        // flushing output but since the headers are likely to be go earlier than content
+        // they don't really conflict so for now I've kept them separate
+        (props.fetchPriority === 'high' ||
+          headers.highImagePreloads.length < 500)
+      ) {
+        // If we postpone in the shell we will still emit this preload so we track
+        // it to make sure we don't reset it.
+        renderState.resets.image[key] = PRELOAD_NO_CREDS;
+        // We manually construct the options for the preload only from strings. We don't want to pollute
+        // the params list with arbitrary props and if we copied everything over as it we might get
+        // coercion errors. We have checks for this in Dev but it seems safer to just only accept values
+        // that are strings
+        const header = getPreloadAsHeader(src, 'image', {
+          imageSrcSet: props.srcSet,
+          imageSizes: props.sizes,
+          crossOrigin,
           integrity: props.integrity,
+          nonce: props.nonce,
           type: props.type,
           fetchPriority: props.fetchPriority,
-          referrerPolicy: props.referrerPolicy,
-        }: PreloadProps),
-      );
-      if (
-        props.fetchPriority === 'high' ||
-        renderState.highImagePreloads.size < 10
-      ) {
-        renderState.highImagePreloads.add(resource);
+          referrerPolicy: props.refererPolicy,
+        });
+        headers.highImagePreloads += header;
+        headers.remainingCapacity -= header.length;
```

Comment:
If this ends up exceeding capacity, can't we just throw it away and and take the other path instead?

So we don't need the hacky "hint" which may or may not be fully honored.

Reply:
yeah, i suppose you only pay the cost once so we can do the limit thing. But per my other comment the limit is still quite imprecise